apps/sync/js/adapters/history.js

```diff
@@ -0,0 +1,309 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+  History Data Adapter
+
+    The feature of History adapter is to retrieve History records from FxSync
+  Service and merge to Places DataStore which has the same functionality, and
+  the adapter is based on SyncEngine's design to handle 'update' and
+  'handleConflict' request. History Data Adapter provides Read-Only version
+  currently, and the basic ideas are included:
+  - Traverse all records from History Collection (HC).
+  - Add a new or existed place to Places DataStore (PDS).
+    - Convert the records from HC to PDS format.
+    - Merge a new record and the existed one then write to PDS.
+  - Remove an existed place when receiving a "deleted: true" record.
+**/
+
+'use strict';
+
+/* global
+  DataAdapters,
+  LazyLoader
+*/
+
+const HISTORY_COLLECTION_MTIME = 'LastSyncedStatus::HistoryCollection::mtime';
+
+var HistoryHelper = (() => {
+  var placesStore;
+  function _ensureStore() {
+    if (placesStore) {
+      return Promise.resolve(placesStore);
+    }
+    return new Promise(resolve => {
+      navigator.getDataStores('places').then(stores => {
+        placesStore = stores[0];
+        resolve(placesStore);
+      });
```

Comment:
No need to wrap this within a new Promise. You can do:
   
    return navigator.getDataStores('places').then(stores => {
       placesStore = stores[0];
       return placesStore;
    });